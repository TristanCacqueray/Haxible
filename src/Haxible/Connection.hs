-- | This module contains the logic to interact with the python wrapper
module Haxible.Connection (Connections (..), TaskCall (..), withConnections) where

import Control.Exception (bracket)
import Data.Aeson (Value (Object), eitherDecodeStrict, encode)
import Data.Aeson.Key qualified
import Data.Aeson.KeyMap qualified
import Data.Bifunctor (first)
import Data.ByteString (hGetLine, toStrict)
import Data.ByteString.Char8 (hPutStrLn)
import Data.Hashable (Hashable)
import Data.Pool qualified
import Data.Text (Text)
import Data.Text qualified as Text
import Data.Typeable (Typeable)
import GHC.Generics (Generic)
import GHC.IO.Handle (hFlush)
import Say
import System.IO (Handle, hClose)
import System.Process.Typed

data TaskCall = TaskCall
  { -- | The playbook attributes, such as `hosts` or `become`
    playAttrs :: [(Text, Value)],
    -- | The task object, e.g `{"file": {"path": "/etc/zuul"}}`
    taskObject :: Value,
    -- | A list of variables, such as `register` names or `include_vars`
    env :: [(Text, Value)]
  }
  deriving (Eq, Show, Typeable, Generic, Hashable)

-- | A connection run converts a TaskCall into a (result code, result value)
newtype Connections = Connections {run :: TaskCall -> IO (Int, Value)}

-- | Creates the Python interpreters.
withConnections :: Int -> FilePath -> (Connections -> IO ()) -> IO ()
withConnections count inventory callback =
  bracket (Data.Pool.newPool poolConfig) Data.Pool.destroyAllResources go
  where
    go pool = do
      putStrLn "Pool ready"

      let runTask :: TaskCall -> Process Handle Handle () -> IO (Int, Value)
          runTask taskCall p = do
            let callParams = [mkObj taskCall.playAttrs, taskCall.taskObject, mkObj taskCall.env]
            say $ " â–¶ Calling " <> Text.pack (show taskCall)
            hPutStrLn (getStdin p) (toStrict $ encode callParams)
            hFlush (getStdin p)
            output <- hGetLine (getStdout p)
            case eitherDecodeStrict output of
              Right res -> pure res
              Left err -> error $ show output <> ": " <> err

      callback (Connections $ Data.Pool.withResource pool . runTask)

    mkObj = Object . Data.Aeson.KeyMap.fromList . map (first Data.Aeson.Key.fromText)

    poolConfig :: Data.Pool.PoolConfig (Process Handle Handle ())
    poolConfig =
      Data.Pool.PoolConfig
        { createResource = do
            startProcess
              . setStdin createPipe
              . setStdout createPipe
              $ proc "python" ["./app/wrapper.py", inventory],
          freeResource = \p -> do
            hClose (getStdin p)
            stopProcess p,
          poolCacheTTL = 3600,
          poolMaxResources = count
        }
