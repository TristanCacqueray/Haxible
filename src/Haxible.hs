-- | The Haxible library entrypoint.
module Haxible where

import Haxible.Codegen
import Haxible.Import
import Haxible.Normalize
import Haxible.Prelude
import Haxible.Syntax
import System.Process.Typed (ExitCode (..), proc, readProcess, runProcess_)

-- | Transform an inventory and playbook into a haskell script.
compile :: FilePath -> FilePath -> IO Text
compile inventory playPath = do
  basePlays <- Haxible.Syntax.decodeFile playPath
  plays <- traverse (Haxible.Import.resolveImport playPath) basePlays
  let exprs = Haxible.Normalize.normalizePlaybook plays
  pure $ renderScript inventory playPath exprs []

-- | Execute the haskell script with cabal.
execute :: FilePath -> IO ()
execute script = runProcess_ (proc "cabal" ["run", "-O0", script])

validate :: FilePath -> IO ()
validate script = do
  (res, stdout, stderr) <- readProcess (proc "runhaskell" [script])
  case res of
    ExitSuccess -> pure ()
    ExitFailure x -> error $ "Process failed with " <> show x <> "\n" <> unsafeFrom stdout <> "\n" <> unsafeFrom stderr
