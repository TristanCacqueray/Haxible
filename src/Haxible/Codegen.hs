module Haxible.Codegen (renderScript) where

import Data.Aeson (encode)
import Data.Aeson.Key qualified
import Data.Aeson.KeyMap qualified
import Data.Text qualified as Text
import Haxible.Normalize
import Haxible.Prelude

renderScript :: FilePath -> [Definition] -> Text
renderScript inventory defs =
  Text.unlines $
    [ "-- Generated with haxible",
      "{-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}",
      "{- cabal:",
      "build-depends: base, haxible",
      "ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T",
      "-}",
      "module Main (main) where\n",
      "import Haxible.Eval\n",
      "main :: IO ()",
      "main = runHaxible " <> quote (from inventory) <> " (playbook [] [])\n"
    ]
      <> concatMap renderDefinition defs

renderDefinition :: Definition -> [Text]
renderDefinition def =
  [ def.name <> " :: Vars -> Vars -> AnsibleHaxl [Value]",
    def.name <> " playAttrs baseEnv = do"
  ]
    <> (mappend "  " <$> map renderExpr def.exprs)
    <> ["  pure $ " <> outputList, ""]
  where
    outputList = Text.intercalate " <> " (toOutput <$> def.exprs)
    toOutput expr = case expr.term of
      -- Module call produces a single value
      ModuleCall _ -> "[" <> from expr.binder <> "]"
      -- Otherwise we got a list of values
      _ -> from expr.binder

renderExpr :: Expr -> Text
renderExpr e = from e.binder <> " <- " <> Text.unwords finalExpr
  where
    requirements = textList $ (\req -> "(" <> quote req.name <> ", " <> mkOrigin req.origin <> ")") <$> e.requirements
    mkOrigin = \case
      Direct n -> from n
      Nested n i -> from n <> " !! " <> from (show i)
      LoopVar -> "__haxible_loop_item"

    finalExpr = case e.loop of
      Just (Array xs) -> mkTraverse $ textList (embedJSON <$> toList xs)
      Just (String v) -> mkTraverse $ "(envLoop " <> quote v <> " " <> paren (requirements <> " <> baseEnv") <> ")"
      Just _ -> error $ "Invalid loop expression: " <> show e.loop
      Nothing -> callExpr
      where
        traverser = case e.term of
          ModuleCall _ -> "traverseLoop"
          DefinitionCall _ -> "traverseInclude"
        mkTraverse arg = [traverser, "(\\__haxible_loop_item -> "] <> callExpr <> [") ", arg]

    callExpr = case e.term of
      ModuleCall CallModule {module_, params, taskAttrs} ->
        [ "runTask",
          "playAttrs",
          quote module_,
          embedJSON (obj $ [(module_, params)] <> taskAttrs),
          paren (requirements <> " <> baseEnv")
        ]
      DefinitionCall CallDefinition {name, playAttrs, baseEnv} ->
        [ name,
          paren (textList (mkJsonArg <$> playAttrs) <> " <> playAttrs"),
          paren (requirements <> " <> " <> textList (mkJsonArg <$> baseEnv) <> " <> baseEnv")
        ]

paren :: Text -> Text
paren = Text.cons '(' . flip Text.snoc ')'

quote :: Text -> Text
quote = Text.cons '"' . flip Text.snoc '"'

embedJSON :: Value -> Text
embedJSON v = "[json|" <> unsafeFrom (Data.Aeson.encode v) <> "|]"

mkJsonArg :: (Text, Value) -> Text
mkJsonArg (n, v) = "(" <> quote n <> ", " <> embedJSON v <> ")"

textList :: [Text] -> Text
textList xs = "[" <> Text.intercalate ", " xs <> "]"

obj :: [(Text, Value)] -> Value
obj = Object . Data.Aeson.KeyMap.fromList . map (first Data.Aeson.Key.fromText)
