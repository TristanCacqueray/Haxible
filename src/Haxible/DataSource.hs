-- | This module contains the Haxl logic
module Haxible.DataSource (AnsibleHaxl, initHaxibleState, dataFetch, TaskReq (..)) where

import Control.Concurrent (QSem, newQSem, signalQSem, waitQSem)
import Control.Concurrent.Async (async)
import Control.Exception (SomeException, bracket_, try)
import Control.Monad (void)
import Data.Aeson (Value)
import Data.Hashable (Hashable (hashWithSalt))
import Data.List qualified as List
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import Data.Text qualified as Text
import Data.Typeable (Typeable)
import Haxible.Connection (Python (..))
import Haxl.Core
import Say

type AnsibleHaxl a = GenHaxl () () a

data TaskReq a where
  RunTask :: Maybe Text -> Text -> Value -> TaskReq Value
  deriving (Typeable)

deriving instance Eq (TaskReq a)

deriving instance Show (TaskReq a)

instance ShowP TaskReq where showp = show

instance Hashable (TaskReq a) where
  hashWithSalt s (RunTask _ n v) = hashWithSalt s (0 :: Int, n, v)

instance StateKey TaskReq where
  data State TaskReq = AnsibleState
    { python :: Python,
      semaphore :: QSem
    }

instance DataSourceName TaskReq where
  dataSourceName _ = "Ansible"

instance DataSource u TaskReq where
  fetch = fetchTask

initHaxibleState :: Int -> Python -> IO (State TaskReq)
initHaxibleState threads python = do
  semaphore <- newQSem threads
  pure $ AnsibleState {python, semaphore}

fetchTask :: State TaskReq -> Flags -> u -> PerformFetch TaskReq
fetchTask state _flags _user =
  BackgroundFetch $ \reqs -> do
    say $ "[+] Batching " <> Text.pack (show (List.length reqs)) <> " tasks"
    mapM_ (fetchAsync state.python state.semaphore) reqs

fetchAsync :: Python -> QSem -> BlockedFetch TaskReq -> IO ()
fetchAsync python sem (BlockedFetch req rvar) =
  void $
    async $
      bracket_ (waitQSem sem) (signalQSem sem) $ do
        e <- Control.Exception.try $ runTaskReq python req
        case e of
          Left ex -> putFailure rvar (ex :: SomeException)
          Right a -> putSuccess rvar a

runTaskReq :: Python -> TaskReq a -> IO a
runTaskReq python (RunTask name action attr) = do
  say $ " â–¶ Calling " <> fromMaybe "unnamed task" name
  python.call action attr
