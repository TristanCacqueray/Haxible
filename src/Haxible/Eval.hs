-- | This module contains the evaluation logic
module Haxible.Eval
  ( AnsibleHaxl,
    runHaxible,
    runTask,
    json,
    envLoop,
    extractFact,
    traverseLoop,
    traverseInclude,
    cleanVar,
    tryRescue,
    Value,
    Vars,
  )
where

import Data.Aeson hiding (json)
import Data.Aeson.Key qualified
import Data.Aeson.KeyMap qualified
import Data.Default (def)
import Data.Functor.Identity (runIdentity)
import Data.Vector qualified
import Haxible.Connection
import Haxible.DataSource
import Haxible.Prelude
import Haxible.Report
import Haxl.Core hiding (env)
import Text.Ginger

type BlockFun = Vars -> Vars -> AnsibleHaxl [Value]

-- $setup
-- >>> let dump = putStrLn . unsafeFrom . encode

-- | Call the rescue block if the main block fails.
tryRescue :: BlockFun -> BlockFun -> BlockFun
tryRescue main rescue a b = do
  res <- tryToHaxlException (main a b)
  case res of
    Left _ -> rescue a b
    Right x -> pure x

-- | Evaluate template for the loop attribute when it is a string (instead of a list):
--
-- >>> envLoop "{{ hosts }}" [("hosts", [json|["frontend", "backend"]|])]
-- [String "frontend",String "backend"]
envLoop :: Text -> [(Text, Value)] -> [Value]
envLoop n env = runGinger context template
  where
    lookupVar var = maybe def toGVal (lookup var env)
    gvarToList g = case g.asList of
      Just xs -> case traverse fromGValEither xs of
        Right x -> x
        Left err -> error $ "Couldn't resolve loop list: " <> err
      Nothing -> error $ "Invalid loop variable: " <> show g

    context = makeContext' lookupVar gvarToList Nothing
    template :: Template SourcePos
    template = case runIdentity (parseGinger (const undefined) Nothing (from n)) of
      Right tmpl -> tmpl
      Left e -> error $ "Template fail: " <> show e

-- | Build the loop results:
--
-- >>> dump =<< traverseLoop (pure . String) ["a", "b"]
-- {"changed":false,"msg":"All items completed","results":["a","b"],"skipped":false}
traverseLoop :: Applicative f => (a -> f Value) -> [a] -> f Value
traverseLoop f xs = loopResult <$> traverse f xs

-- | Concat nested results
--
-- >>> dump =<< traverseInclude (\x -> pure [String x, Number 42]) ["a", "b"]
-- ["a",42,"b",42]
traverseInclude :: Applicative f => (a -> f [Value]) -> [a] -> f [Value]
traverseInclude f xs = concat <$> traverse f xs

-- | Extract the ansible_facts results
--
-- >>> dump $ extractFact [json|{"ansible_facts": {"key": "value"}}|]
-- "value"
--
-- Each fact are evaluated individually, so multiple result is un-expected:
--
-- >>> extractFacts [json|{"ansible_facts": {"x": 1, "y": 2}}|]
-- Left "Multiple facts found: [(\"x\",Number 1.0),(\"y\",Number 2.0)]"
extractFacts :: Value -> Either String Value
extractFacts v = case preview (key "ansible_facts") v of
  Just (Object obj) -> case Data.Aeson.KeyMap.toList obj of
    [(_, value)] -> Right value
    xs -> Left $ "Multiple facts found: " <> show xs
  _ -> case preview (key "skip_reason" . _String) v of
    Just _ -> Right Null
    Nothing -> Left $ "Can't find ansible_facts in " <> unsafeFrom (encode v)

extractFact :: Value -> Value
extractFact v = case extractFacts v of
  Left e -> error e
  Right x -> x

-- | Process loop results
--
-- >>> dump $ loopResult [[json|42|], [json|43|]]
-- {"changed":false,"msg":"All items completed","results":[42,43],"skipped":false}
loopResult :: [Value] -> Value
loopResult xs = Object $ Data.Aeson.KeyMap.fromList attrs
  where
    play = case xs of
      (x : _) -> copyKey "__haxible_play" x <> copyKey "__haxible_module" x
      _ -> []
    copyKey k x = maybe [] (\p -> [(Data.Aeson.Key.fromText k, p)]) (preview (key k) x)
    attrs =
      [ ("results", Array (Data.Vector.fromList xs)),
        ("msg", "All items completed"),
        ("skipped", Bool $ all (fromMaybe False . preview (key "skipped" . _Bool)) xs),
        ("changed", Bool $ any (fromMaybe False . preview (key "changed" . _Bool)) xs)
      ]
        <> play

runTask :: Vars -> Text -> Value -> Vars -> Vars -> AnsibleHaxl Value
runTask playAttrs module_ moduleObject taskAttrs baseTaskVars =
  addModule <$> dataFetch (RunTask (TaskCall {playAttrs, moduleObject, taskAttrs, taskVars, module_}))
  where
    addModule = \case
      Object obj -> Object $ Data.Aeson.KeyMap.insert "__haxible_module" (String module_) obj
      x -> x
    taskVars = filter ((/=) Null . snd) (concatMap checkManyHost baseTaskVars)
    -- When a task run on many host, we register a single variable with all the results,
    -- thus when accessing the variable, we need to lookup the current host result.
    checkManyHost (k, v)
      | preview (key "__haxible_multi_hosts" . _Bool) v == Just True =
          let many_k = k <> "__haxible"
           in [ (k, String $ "{{ " <> many_k <> "[ansible_host] }}"),
                (many_k, cleanVar v)
              ]
      | otherwise = [(k, cleanVar v)]

runHaxible :: FilePath -> FilePath -> AnsibleHaxl [Value] -> IO ()
runHaxible inventory playPath action = withConnections 5 inventory $ \connections -> do
  ansibleState <- initHaxibleState connections
  ansibleEnv <- initEnv (stateSet ansibleState stateEmpty) ()
  xs <- runHaxl ansibleEnv action
  traverse_ (putStrLn . unsafeFrom . encode) xs
  case nonEmpty xs of
    Just res -> do
      writeFile (playName <> ".plantuml") (from $ Haxible.Report.reportTiming res)
    Nothing -> putStrLn "\nEmpty results :("
  where
    (playName, _) = splitExtension playPath
