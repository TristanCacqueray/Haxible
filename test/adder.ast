[ Definition
    { name = "playbook"
    , requires = []
    , provides = []
    , outputs = Environment
        { getEnv = [] }
    , exprs =
        [ Expr
            { binder = Binder "resultsLocalhost0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "debug0"
                    , dep = Register "answer"
                    }
                ]
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "playLocalhost0"
                    , playAttrs =
                        [
                            ( "gather_facts"
                            , Bool False
                            )
                        ,
                            ( "hosts"
                            , String "localhost"
                            )
                        ]
                    , baseEnv = []
                    }
                )
            }
        , Expr
            { binder = Binder "resultsBackend0"
            , requires =
                [ Resource
                    { name = Binder "debug0"
                    , dep = Register "answer"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "answer"
                    , origin = Nested
                        ( Binder "resultsLocalhost0" ) 0
                    }
                ]
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "playBackend0"
                    , playAttrs =
                        [
                            ( "gather_facts"
                            , Bool False
                            )
                        ,
                            ( "hosts"
                            , String "backend"
                            )
                        ]
                    , baseEnv = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "playLocalhost0"
    , requires = []
    , provides =
        [ Resource
            { name = Binder "debug0"
            , dep = Register "answer"
            }
        ]
    , outputs = Environment
        { getEnv =
            [
                ( Binder "debug0"
                , Right
                    [ Resource
                        { name = Binder "debug0"
                        , dep = Register "answer"
                        }
                    ]
                )
            ,
                ( Binder "assert0"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "debug0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "debug0"
                    , dep = Register "answer"
                    }
                ]
            , outputs = Right
                [ Resource
                    { name = Binder "debug0"
                    , dep = Register "answer"
                    }
                ]
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "debug"
                    , params = Object
                        ( fromList
                            [
                                ( "msg"
                                , String "42"
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        , Expr
            { binder = Binder "assert0"
            , requires =
                [ Resource
                    { name = Binder "debug0"
                    , dep = Register "answer"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "answer"
                    , origin = Direct
                        ( Binder "debug0" )
                    }
                ]
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "assert"
                    , params = Object
                        ( fromList
                            [
                                ( "that"
                                , Array
                                    [ String "answer['msg'] == '42'" ]
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "playBackend0"
    , requires =
        [ Resource
            { name = Binder "debug0"
            , dep = Register "answer"
            }
        ]
    , provides = []
    , outputs = Environment
        { getEnv =
            [
                ( Binder "roleAdder0"
                , Right []
                )
            ,
                ( Binder "debug1"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "roleAdder0"
            , requires =
                [ Resource
                    { name = Binder "debug0"
                    , dep = Register "answer"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "roleAdder"
                    , playAttrs = []
                    , baseEnv =
                        [
                            ( "x"
                            , String "{{ answer['msg'] }}"
                            )
                        ,
                            ( "y"
                            , String "21"
                            )
                        ]
                    }
                )
            }
        , Expr
            { binder = Binder "debug1"
            , requires = []
            , provides = []
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "debug"
                    , params = Object
                        ( fromList
                            [
                                ( "msg"
                                , String "Over!"
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "roleAdder"
    , requires = []
    , provides =
        [ Resource
            { name = Binder "debugAddingNumbers0"
            , dep = Register "add_result"
            }
        ]
    , outputs = Environment
        { getEnv =
            [
                ( Binder "debugAddingNumbers0"
                , Right
                    [ Resource
                        { name = Binder "debugAddingNumbers0"
                        , dep = Register "add_result"
                        }
                    ]
                )
            ,
                ( Binder "assertCheckingResults0"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "debugAddingNumbers0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "debugAddingNumbers0"
                    , dep = Register "add_result"
                    }
                ]
            , outputs = Right
                [ Resource
                    { name = Binder "debugAddingNumbers0"
                    , dep = Register "add_result"
                    }
                ]
            , requirements = []
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "debug"
                    , params = Object
                        ( fromList
                            [
                                ( "msg"
                                , String "Adding {{ x }} + {{ y }}"
                                )
                            ]
                        )
                    , taskAttrs =
                        [
                            ( "name"
                            , String "Adding numbers"
                            )
                        ]
                    }
                )
            }
        , Expr
            { binder = Binder "assertCheckingResults0"
            , requires =
                [ Resource
                    { name = Binder "debugAddingNumbers0"
                    , dep = Register "add_result"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "add_result"
                    , origin = Direct
                        ( Binder "debugAddingNumbers0" )
                    }
                ]
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "assert"
                    , params = Object
                        ( fromList
                            [
                                ( "that"
                                , Array
                                    [ String "x == '42' and y == '21'"
                                    , String "add_result['msg'] == 'Adding 42 + 21'"
                                    ]
                                )
                            ]
                        )
                    , taskAttrs =
                        [
                            ( "name"
                            , String "Checking results"
                            )
                        ]
                    }
                )
            }
        ]
    }
]