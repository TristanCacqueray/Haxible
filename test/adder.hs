-- Generated with haxible
{-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}
{- cabal:
build-depends: base, haxible
ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T
-}
module Main (main) where

import Haxible.Eval

main :: IO ()
main = runHaxible "inventory.yaml" (playbook [] [])

playbook :: Vars -> Vars -> AnsibleHaxl [Value]
playbook playAttrs baseEnv = do
  resultsLocalhost0 <- playLocalhost0 ([("gather_facts", [json|false|]), ("hosts", [json|"localhost"|])] <> playAttrs) ([] <> [] <> baseEnv)
  resultsBackend0 <- playBackend0 ([("gather_facts", [json|false|]), ("hosts", [json|"backend"|])] <> playAttrs) ([("answer", resultsLocalhost0 !! 0)] <> [] <> baseEnv)
  pure $ resultsLocalhost0 <> resultsBackend0

playLocalhost0 :: Vars -> Vars -> AnsibleHaxl [Value]
playLocalhost0 playAttrs baseEnv = do
  debug0 <- runTask playAttrs "debug" [json|{"debug":{"msg":"42"}}|] ([] <> baseEnv)
  assert0 <- runTask playAttrs "assert" [json|{"assert":{"that":["answer['msg'] == '42'"]}}|] ([("answer", debug0)] <> baseEnv)
  pure $ [debug0] <> [assert0]

playBackend0 :: Vars -> Vars -> AnsibleHaxl [Value]
playBackend0 playAttrs baseEnv = do
  roleAdder0 <- roleAdder ([] <> playAttrs) ([] <> [("x", [json|"{{ answer['msg'] }}"|]), ("y", [json|"21"|])] <> baseEnv)
  debug1 <- runTask playAttrs "debug" [json|{"debug":{"msg":"Over!"}}|] ([] <> baseEnv)
  pure $ roleAdder0 <> [debug1]

roleAdder :: Vars -> Vars -> AnsibleHaxl [Value]
roleAdder playAttrs baseEnv = do
  debugAddingNumbers0 <- runTask playAttrs "debug" [json|{"debug":{"msg":"Adding {{ x }} + {{ y }}"},"name":"Adding numbers"}|] ([] <> baseEnv)
  assertCheckingResults0 <- runTask playAttrs "assert" [json|{"assert":{"that":["x == '42' and y == '21'","add_result['msg'] == 'Adding 42 + 21'"]},"name":"Checking results"}|] ([("add_result", debugAddingNumbers0)] <> baseEnv)
  pure $ [debugAddingNumbers0] <> [assertCheckingResults0]

