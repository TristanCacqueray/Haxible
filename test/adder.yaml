- hosts: localhost
  gather_facts: no
  tasks:
    - debug:
        msg: "42"
      register: answer

    - assert:
        that:
          - "answer['msg'] == '42'"

- hosts: backend
  gather_facts: no
  tasks:
    - name: Calling adder
      include_role:
        name: adder
      vars:
        x: "{{ answer['msg'] }}"
        y: "21"
