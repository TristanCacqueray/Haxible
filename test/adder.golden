(
    [ Definition
        { name = "playbook"
        , requires = []
        , provides = []
        , outputs = Environment
            { getEnv = [] }
        , exprs =
            [ Expr
                { binder = Binder "resultsLocalhost0"
                , requires = []
                , provides =
                    [ Resource
                        { name = Binder "debug0"
                        , dep = Register "answer"
                        }
                    ]
                , outputs = Right []
                , requirements = []
                , loop = Nothing
                , term = DefinitionCall
                    ( CallDefinition
                        { name = "playLocalhost0"
                        , playAttrs =
                            [
                                ( "gather_facts"
                                , Bool False
                                )
                            ,
                                ( "hosts"
                                , String "localhost"
                                )
                            ]
                        , baseEnv = []
                        }
                    )
                }
            , Expr
                { binder = Binder "resultsBackend0"
                , requires =
                    [ Resource
                        { name = Binder "debug0"
                        , dep = Register "answer"
                        }
                    ]
                , provides = []
                , outputs = Right []
                , requirements =
                    [ Requirement
                        { name = "answer"
                        , origin = Nested
                            ( Binder "resultsLocalhost0" ) 0
                        }
                    ]
                , loop = Nothing
                , term = DefinitionCall
                    ( CallDefinition
                        { name = "playBackend0"
                        , playAttrs =
                            [
                                ( "gather_facts"
                                , Bool False
                                )
                            ,
                                ( "hosts"
                                , String "backend"
                                )
                            ]
                        , baseEnv = []
                        }
                    )
                }
            ]
        }
    , Definition
        { name = "playLocalhost0"
        , requires = []
        , provides =
            [ Resource
                { name = Binder "debug0"
                , dep = Register "answer"
                }
            ]
        , outputs = Environment
            { getEnv =
                [
                    ( Binder "debug0"
                    , Right
                        [ Resource
                            { name = Binder "debug0"
                            , dep = Register "answer"
                            }
                        ]
                    )
                ,
                    ( Binder "assert0"
                    , Right []
                    )
                ]
            }
        , exprs =
            [ Expr
                { binder = Binder "debug0"
                , requires = []
                , provides =
                    [ Resource
                        { name = Binder "debug0"
                        , dep = Register "answer"
                        }
                    ]
                , outputs = Right
                    [ Resource
                        { name = Binder "debug0"
                        , dep = Register "answer"
                        }
                    ]
                , requirements = []
                , loop = Nothing
                , term = ModuleCall
                    ( CallModule
                        { module_ = "debug"
                        , params = Object
                            ( fromList
                                [
                                    ( "msg"
                                    , String "42"
                                    )
                                ]
                            )
                        , taskAttrs = []
                        }
                    )
                }
            , Expr
                { binder = Binder "assert0"
                , requires =
                    [ Resource
                        { name = Binder "debug0"
                        , dep = Register "answer"
                        }
                    ]
                , provides = []
                , outputs = Right []
                , requirements =
                    [ Requirement
                        { name = "answer"
                        , origin = Direct
                            ( Binder "debug0" )
                        }
                    ]
                , loop = Nothing
                , term = ModuleCall
                    ( CallModule
                        { module_ = "assert"
                        , params = Object
                            ( fromList
                                [
                                    ( "that"
                                    , Array
                                        [ String "answer['msg'] == '42'" ]
                                    )
                                ]
                            )
                        , taskAttrs = []
                        }
                    )
                }
            ]
        }
    , Definition
        { name = "playBackend0"
        , requires =
            [ Resource
                { name = Binder "debug0"
                , dep = Register "answer"
                }
            ]
        , provides = []
        , outputs = Environment
            { getEnv =
                [
                    ( Binder "roleAdder0"
                    , Right []
                    )
                ]
            }
        , exprs =
            [ Expr
                { binder = Binder "roleAdder0"
                , requires =
                    [ Resource
                        { name = Binder "debug0"
                        , dep = Register "answer"
                        }
                    ]
                , provides = []
                , outputs = Right []
                , requirements = []
                , loop = Nothing
                , term = DefinitionCall
                    ( CallDefinition
                        { name = "roleAdder"
                        , playAttrs = []
                        , baseEnv =
                            [
                                ( "x"
                                , String "{{ answer['msg'] }}"
                                )
                            ,
                                ( "y"
                                , String "21"
                                )
                            ]
                        }
                    )
                }
            ]
        }
    , Definition
        { name = "roleAdder"
        , requires = []
        , provides =
            [ Resource
                { name = Binder "debug1"
                , dep = Register "add_result"
                }
            ]
        , outputs = Environment
            { getEnv =
                [
                    ( Binder "debug1"
                    , Right
                        [ Resource
                            { name = Binder "debug1"
                            , dep = Register "add_result"
                            }
                        ]
                    )
                ,
                    ( Binder "assert1"
                    , Right []
                    )
                ]
            }
        , exprs =
            [ Expr
                { binder = Binder "debug1"
                , requires = []
                , provides =
                    [ Resource
                        { name = Binder "debug1"
                        , dep = Register "add_result"
                        }
                    ]
                , outputs = Right
                    [ Resource
                        { name = Binder "debug1"
                        , dep = Register "add_result"
                        }
                    ]
                , requirements = []
                , loop = Nothing
                , term = ModuleCall
                    ( CallModule
                        { module_ = "debug"
                        , params = Object
                            ( fromList
                                [
                                    ( "msg"
                                    , String "Adding {{ x }} + {{ y }}"
                                    )
                                ]
                            )
                        , taskAttrs = []
                        }
                    )
                }
            , Expr
                { binder = Binder "assert1"
                , requires =
                    [ Resource
                        { name = Binder "debug1"
                        , dep = Register "add_result"
                        }
                    ]
                , provides = []
                , outputs = Right []
                , requirements =
                    [ Requirement
                        { name = "add_result"
                        , origin = Direct
                            ( Binder "debug1" )
                        }
                    ]
                , loop = Nothing
                , term = ModuleCall
                    ( CallModule
                        { module_ = "assert"
                        , params = Object
                            ( fromList
                                [
                                    ( "that"
                                    , Array
                                        [ String "x == '42' and y == '21'"
                                        , String "add_result['msg'] == 'Adding 42 + 21'"
                                        ]
                                    )
                                ]
                            )
                        , taskAttrs = []
                        }
                    )
                }
            ]
        }
    ]
, "-- Generated with haxible
  {-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}
  {- cabal:
  build-depends: base, haxible
  ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T
  -}
  module Main (main) where

  import Haxible.Eval

  main :: IO ()
  main = runHaxible "inventory.yaml" (playbook [] [])

  playbook :: Vars -> Vars -> AnsibleHaxl [Value]
  playbook playAttrs baseEnv = do
    resultsLocalhost0 <- playLocalhost0 ([("gather_facts", [json|false|]), ("hosts", [json|"localhost"|])] <> playAttrs) ([] <> [] <> baseEnv)
    resultsBackend0 <- playBackend0 ([("gather_facts", [json|false|]), ("hosts", [json|"backend"|])] <> playAttrs) ([("answer", resultsLocalhost0 !! 0)] <> [] <> baseEnv)
    pure $ resultsLocalhost0 <> resultsBackend0

  playLocalhost0 :: Vars -> Vars -> AnsibleHaxl [Value]
  playLocalhost0 playAttrs baseEnv = do
    debug0 <- runTask playAttrs [json|{"debug":{"msg":"42"}}|] ([] <> baseEnv)
    assert0 <- runTask playAttrs [json|{"assert":{"that":["answer['msg'] == '42'"]}}|] ([("answer", debug0)] <> baseEnv)
    pure $ [debug0] <> [assert0]

  playBackend0 :: Vars -> Vars -> AnsibleHaxl [Value]
  playBackend0 playAttrs baseEnv = do
    roleAdder0 <- roleAdder ([] <> playAttrs) ([] <> [("x", [json|"{{ answer['msg'] }}"|]), ("y", [json|"21"|])] <> baseEnv)
    pure $ roleAdder0

  roleAdder :: Vars -> Vars -> AnsibleHaxl [Value]
  roleAdder playAttrs baseEnv = do
    debug1 <- runTask playAttrs [json|{"debug":{"msg":"Adding {{ x }} + {{ y }}"}}|] ([] <> baseEnv)
    assert1 <- runTask playAttrs [json|{"assert":{"that":["x == '42' and y == '21'","add_result['msg'] == 'Adding 42 + 21'"]}}|] ([("add_result", debug1)] <> baseEnv)
    pure $ [debug1] <> [assert1]

  "
)